cpp/ql/src/Likely Bugs/OO/UnsafeUseOfThis.ql

```diff
@@ -0,0 +1,288 @@
+        succ.(CopyInstruction).getSourceValue() = instr or
+        succ.(CheckedConvertOrNullInstruction).getUnary() = instr or
+        succ.(ChiInstruction).getTotal() = instr or
+        succ.(ConvertInstruction).getUnary() = instr
+      )
+      or
+      succ.(InheritanceConversionInstruction).getUnary() = instr
+    )
+    or
+    flowThroughCallable(instr, succ, sinkClass)
+  )
+}
+
+/**
+ * Holds if
+ * - `instr` is an argument (or argument indirection) to a call, and
+ * - `succ` is the corresponding initialization instruction in the call target, and
+ * - there exists an initialization of a `this` pointer of type `sourceClass` that flows to `instr`.
+ */
+predicate flowThroughCallableFwd(Instruction instr, Instruction succ, Class sourceClass) {
+  flowsFromSource(_, instr, sourceClass) and
+  (
+    // Flow from an argument to a parameter
+    exists(CallInstruction call, InitializeParameterInstruction init | init = succ |
+      getPositionalArgumentInitParam(call, instr, init, call.getStaticCallTarget())
+      or
+      getThisArgumentInitParam(call, instr, init, call.getStaticCallTarget())
+    )
+    or
+    // Flow from argument indirection to parameter indirection
+    exists(
+      CallInstruction call, ReadSideEffectInstruction read, InitializeIndirectionInstruction init
+    |
+      init = succ and
+      init.getEnclosingFunction() = call.getStaticCallTarget() and
+      read.getPrimaryInstruction() = call and
+      read.getSideEffectOperand().getAnyDef() = instr
+    |
+      exists(int n |
+        read.getIndex() = n and
+        init.getParameter().getIndex() = unbind(n)
+      )
+      or
+      call.getThisArgument() = instr and
+      init.getIRVariable() instanceof IRThisVariable
+    )
+  )
+}
+
+/**
+ * Holds if there exists an initialization of a `this` pointer of type `sourceClass` that flows
+ * to `instr`, and `instr` flows to `succ` and `succ`.
+ */
+predicate successorFwd(Instruction instr, Instruction succ, Class sourceClass) {
+  flowsFromSource(_, instr, sourceClass) and
+  (
+    irBbPostDominates(succ.getBlock(), instr.getBlock()) and
+    (
+      (
+        succ.(CopyInstruction).getSourceValue() = instr or
+        succ.(CheckedConvertOrNullInstruction).getUnary() = instr or
+        succ.(ChiInstruction).getTotal() = instr or
+        succ.(ConvertInstruction).getUnary() = instr
+      )
+      or
+      succ.(InheritanceConversionInstruction).getUnary() = instr
+    )
+    or
+    flowThroughCallableFwd(instr, succ, sourceClass)
+  )
+}
+
+/**
+ * Holds if `instr` is in the path from an initialization of a `this` pointer in a subclass, to a use
+ * of the `this` pointer in a baseclass.
+ */
+predicate isInPath(Instruction instr) {
+  exists(Class sourceClass, Class sinkClass |
+    flowsFromSource(_, instr, sourceClass) and
+    flowsToSink(instr, _, sinkClass) and
+    sourceClass.getABaseClass+() = sinkClass
+  )
+}
+
+query predicate edges(Instruction a, Instruction b) { successor(a, b, _) }
```

Comment:
Shouldn't we also check `isInPath` for both `a` and `b` here to avoid producing too many edges?

Reply:
I think so, yes. I've done this in e1ef41bd38af5924d0a4f3cab1267f75f77218e2. It wasn't so bad since `successor` already filters out the ones that flow to a sink, but your addition improves this pruning even more.